L1Trigger/L1THGCal/src/be_algorithms/HGCalClusteringImpl.cc

```diff
@@ -270,3 +276,162 @@ void HGCalClusteringImpl::clusterizeNN( const edm::PtrVector<l1t::HGCalTriggerCe
+        
+        double seedThreshold = ((*tc)->subdetId()==HGCHEB ? scintillatorSeedThreshold_ : siliconSeedThreshold_);
+
+        /* decide if is a seed, if yes store the position into of triggerCellsPtrs */
+        isSeed[itc]       = ( (*tc)->mipPt() > seedThreshold) ? true : false;
+        if( isSeed[itc] ) seedPositions.push_back( itc );
+
+        /* remove tc from the seed vector if is a NN of an other seed*/
+        for( auto pos : seedPositions ){
+            if( ( (*tc)->position() - triggerCellsPtrs[pos]->position() ).mag()  < dr_ ){
+                if( this->areTCneighbour_( (*tc)->detId(), triggerCellsPtrs[pos]->detId(), triggerGeometry ) )
+                {
+                    isSeed[itc] = false;
+                    seedPositions.pop_back();
+                }
+            }
+        } 
+
+    }
+    
+    /* clustering the TCs */
+    std::vector<l1t::HGCalCluster> clustersTmp;
+
+    // every seed generates a cluster
+    for( auto pos : seedPositions ) {
+        clustersTmp.emplace_back( triggerCellsPtrs[pos] );
+    }
+
+    /* add the tc to the clusters */
+    itc=0;
+    for( edm::PtrVector<l1t::HGCalTriggerCell>::const_iterator tc = triggerCellsPtrs.begin(); tc != triggerCellsPtrs.end(); ++tc,++itc ){
+      
+        /* get the correct threshold for the different part of the detector */ 
+        double threshold = ((*tc)->subdetId()==HGCHEB ? scintillatorTriggerCellThreshold_ : siliconTriggerCellThreshold_);
+        
+        /* continue if not passing the threshold */
+        if( (*tc)->mipPt() < threshold ) continue;
+        
+        /* searching for TC near the centre of the cluster  */
+        std::vector<unsigned> tcPertinentClusters; 
+        unsigned iclu(0);
+        
+        for ( const auto& clu : clustersTmp ) {
+            if( this->isPertinent(**tc, clu, dr_) ){
+                tcPertinentClusters.push_back( iclu );
+            }
+            ++iclu;
+        }   
+
+        if ( tcPertinentClusters.size() == 0 ) {
+            continue;
+        }
+        else if( tcPertinentClusters.size() == 1 ) {
+            clustersTmp.at( tcPertinentClusters.at(0) ).addConstituent( *tc );
+        }
+        else {
+
+            /* calculate the fractions */
+            double totMipt = 0;
+            for( auto clu : tcPertinentClusters ){
+                totMipt += clustersTmp.at( clu ).constituents()[0]->mipPt();
+            }
+
+            for( auto clu : tcPertinentClusters ){
+                double seedMipt = clustersTmp.at( clu ).constituents()[0]->mipPt();
+                clustersTmp.at( clu ).addConstituent( *tc, true, seedMipt/totMipt );
+            }
+        }
+    }
+
+    /* store clusters in the persistent collection */
+    clusters.resize(0, clustersTmp.size());
+    for( unsigned i(0); i<clustersTmp.size(); ++i ){
+        this->removeUnconnectedTCinCluster_( &clustersTmp.at(i), triggerGeometry );
```

Comment:
I would rather take a reference to the cluster instead of a pointer in `removeUnconnectedTCinCluster()`. This way there is no need to take the address here. And this is also more consistent with everywhere else in the package.
```cpp
removeUnconnectedTCinCluster_( clustersTmp.at(i), triggerGeometry );
```
and 
```cpp
removeUnconnectedTCinCluster_( l1t::HGCalCluster& cluster, const HGCalTriggerGeometryBase & triggerGeometry )
```